### DISCUSSION TITLE
Add helper function for plugins to detect if other plugins are installed, enabled and a class exists

### DISCUSSION THREAD

**Discussion by engram-design** [craftcms/cms#8967] [Ideas]
For instance, I want another plugin to be able to support Feed Me. Currently, I need to do:

```
use craft\feedme\events\RegisterFeedMeElementsEvent;
use craft\feedme\services\Elements as FeedMeElements;

$isInstalled = Craft::$app->getPlugins()->isPluginInstalled('feed-me') && Craft::$app->getPlugins()->isPluginEnabled('feed-me');
$classExists = class_exists(FeedMeElements::class);

if ($isInstalled && $classExists) {
    Event::on(FeedMeElements::class, FeedMeElements::EVENT_REGISTER_FEED_ME_ELEMENTS, function(RegisterFeedMeElementsEvent $e) {
        // ...
    });
}
```

Firstly, the check for whether the plugin is installed and enabled. From what I can tell, I maybe only need a single check here for if enabled (if a plugin is enabled, it's inferred it's installed). But then the case of whether the class I need is actually available. Maybe they've done a `composer remove craftcms/feed-me` or something similar without uninstalling (in fact, an easy way to reproduce this).

There's an existing helper function https://github.com/craftcms/cms/blob/develop/src/helpers/Component.php#L35 which is very close to this, but only checks if the class provided is the actual class of the plugin, not a plugin-owned class.

Maybe it's creeping into non-framework code, but we use it pretty frequently to check on this sort of thing in plugins.

**Comment by brandonkelly** [comment]
I’m sure there are valid use cases for such a method, but at least in this case, your code could be simplified to just the `class_exists()` check:

```php
if (class_exists(FeedMeElements::class)) {
    Event::on(FeedMeElements::class, FeedMeElements::EVENT_REGISTER_FEED_ME_ELEMENTS, function(RegisterFeedMeElementsEvent $e) {
        // ...
    });
}
```

The `FeedMeElements::EVENT_REGISTER_FEED_ME_ELEMENTS` reference will cause `FeedMeElements.php` to be loaded and its class to be declared, but it won’t cause any of the plugin’s actual code to be run (unless the file happened to have some logic outside of the class definition, which is unlikely), so not a huge deal if the plugin is Composer-installed but not Craft-installed/enabled.

**Comment by engram-design** [comment]
Valid point, I suppose my example isn't ideal, but I swear I've come across an appropriate scenario before, but struggling to find it right now. Do with this issue what you will then!
